File: ios/SwiftKotlination/SwiftKotlination/Sources/Features/Story/ViewModel/StoryViewModel.swift
import UIKit.UIImage

final class StoryViewModel {

    typealias Factory = ImageFactory & StoryFactory
    private let factory: Factory
    private let story: Story

    private lazy var storyManager = factory.makeStoryManager(for: story)
    private lazy var imageManager = factory.makeImageManager()

    init(factory: Factory, story: Story) {
        self.factory = factory
        self.story = story
    }

    var image: UIImage?

    func story(_ observer: @escaping Observer<Story>) {
        storyManager.story(observer)
    }

    @discardableResult
    func image(with url: String, _ observer: @escaping Observer<UIImage>) -> Disposable? {
        guard let image = image else {
            return imageManager
                .image(with: url) { [weak self] result in
                    switch result {
                    case .success(let image):
                        self?.image = image

                    case .failure:
                        break
                    }
                    observer(result)
            }
        }

        observer(.success(image))
        return nil
    }
}

File: android/SwiftKotlination/features/story/src/main/kotlin/fr/jhandguy/story/viewmodel/StoryViewModel.kt
package fr.jhandguy.story.viewmodel

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import fr.jhandguy.network.model.observer.Disposable
import fr.jhandguy.network.model.observer.Observer
import fr.jhandguy.network.model.observer.Result
import fr.jhandguy.story.factory.StoryFactory
import fr.jhandguy.story.model.Story

class StoryViewModel(factory: StoryFactory, story: Story = Story(), var image: Bitmap? = null) {
    private val storyManager = factory.makeStoryManager(story)
    private val imageManager = factory.makeImageManager()

    suspend fun story(observer: Observer<Story>) = storyManager.story(observer)

    suspend fun image(url: String, observer: Observer<Bitmap>): Disposable? = image?.let { image ->
        observer(Result.Success(image))
        return null
    } ?: imageManager.image(url) { result ->
        when (result) {
            is Result.Success -> {
                val image = BitmapFactory.decodeByteArray(result.data, 0, result.data.size)
                this.image = image
                observer(Result.Success(image))
            }
            is Result.Failure -> observer(result)
        }
    }
}
